### PR TITLE
📖 Etcd on data disk CAEP

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What this PR does / why we need it**: 
This PR brings the CAEP for implementing disk setup and mounts in CABPK.

Comments on the [Google doc](https://docs.google.com/document/d/12eYcQrlZdkojHY3atUCCIS9YGPQ1LSQvT4gdg6JWroc/edit?usp=sharing) are addressed.

Moving here for getting feedback from a broader audience.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Relates to #2994

### CODE REVIEW

**Comment by vincepri** [comment]
1 week timeout starting today?

**Comment by vincepri** [comment]
@michaelgugino You put in words something that myself (and probably others) have been feeling for a while. I do think the bootstrap abstraction needs a complete rework, although without time and people-hours dedicated to it we won't be able to deliver such a rework (which will probably also require one or more API revisions to fully become concrete).

Yesterday, @CecileRobertMichon and I were talking about this proposal, and how the bootstrap provider is becoming more and more infrastructure aware and crossing boundaries.

While I do think there are better, longer term solutions, I'm fully behind having these changes go in today because it's an iterative improvement and it opens new cases and features. In addition, we have the team at Microsoft behind this proposal, which is also going to help with implementation, support, and testing.

**Comment by CecileRobertMichon** [comment]
> This goes back to the issue of the problem being a higher-level problem. One can't have an additional disk in the bootstrap config without having the corresponding machine infrastructure in place. So, there still needs to be yet-another layer on top to coordinate these pieces. Is there going to be some higher-level tuneable such as 'give-etcd-separate-disks' that is going to stamp out all the options for all the pieces down below? Doesn't seem feasible. At some point, we have to give users the power to make their own decisions.

@michaelgugino that is indeed a problem I've been thinking about a lot with this proposal, and I also put it in the proposal as the last "risk and mitigation". Right now, the solution I'm proposing is actually give the users the choice ( I personally see it as "putting the burden on the user" though) by exposing a generic disk setup and making the changes required to put etcd on an external disk part of the user template (ie. completely user configurable). There will _not_ be "a higher-level tuneable such as 'give-etcd-separate-disks'". See my capz prototype for an example of this works in practice https://github.com/kubernetes-sigs/cluster-api-provider-azure/compare/master...CecileRobertMichon:etcd-disk#diff-a13d5b9325a3022286a0a671bc66b2cbR113.

I agree that "KubeadmConfigSpec" is taking on too much infrastructure configuration right now but as Vince said, reworking CABPK needs to be its own proposal and would likely involve breaking changes.  I'm trying to unblock CAPZ and CAPA users by providing a generic, reusable implementation that brings incremental value and can be done in v1alpha3.

**Comment by fabriziopandini** [comment]
+1 for me because this enables new use cases right now in v0.3.x. 
I also agree that long term, we should find a way for moving infrastructure specific bits away from KubeadmConfig

**Comment by michaelgugino** [comment]
I don't have strong objection to this particular feature if we've identified that long term this provider is probably not the ideal implementation.

Perhaps, then, this CAEP should be more generic.  Something like "Continue adding cloud-init functionality to CAPBK" so people can add features as they need them instead of having to file a new CAEP every time we want to extend the missing pieces.  Perhaps some framework for expected implementation conformance.

**Comment by dhawal55** [comment]
+1

**Comment by CecileRobertMichon** [comment]
@michaelgugino let's continue the conversation about making cloud init more accessible to the user and infrastructure bootstrap less tied to kubeadm bootstrapper in #3064

File: docs/proposals/20200423-etcd-data-disk.md [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+```
+
+### Instrument the OS image with image-builder to perform this customization automatically through custom UDEV rules, scripts, etc.
+
+## Upgrade Strategy
+
+N/A as this proposal only adds new types.
+
+<!-- Links -->
+[community meeting]: https://docs.google.com/document/d/1Ys-DOR5UsgbMEeciuG0HOgDQc8kZsaWIWJeKJ1-UfbY
```
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply docs/proposals/20200423-etcd-data-disk.md]: removed, the link wasn't being used.

**Comment by vincepri** [comment]
@CecileRobertMichon might want to try to rebase?

**Comment by CecileRobertMichon** [comment]
@vincepri I just did earlier, I'm pretty sure the failures are due to https://status.quay.io/

**Comment by vincepri** [comment]
uh oh! You're right, apologies

**Pull request merged**